base/headers.go

```diff
@@ -227,17 +233,115 @@ func (uri WildcardUri) Equals(other Uri) bool {
 }
 
 // Generic list of parameters on a header.
-type Params map[string]MaybeString
+type Params struct {
+	params     map[string]MaybeString
+	paramOrder []string
+}
+
+// Create an empty set of parameters.
+func NewParams() *Params {
+	return &Params{map[string]MaybeString{}, []string{}}
+}
+
+// Returns the entire parameter map.
+func (p *Params) Params() map[string]MaybeString {
+	return p.params
+}
+
+// Returns the requested parameter value.
+func (p *Params) Get(k string) (MaybeString, bool) {
+	v, ok := p.params[k]
+	return v, ok
+}
+
+func (p *Params) Add(k string, v MaybeString) *Params {
+	if p.params == nil {
+		p.params = map[string]MaybeString{}
+		p.paramOrder = []string{}
+	}
+
+	// Add param to order list if new.
+	if _, ok := p.params[k]; !ok {
+		p.paramOrder = append(p.paramOrder, k)
+	}
+
+	// Set param value.
+	p.params[k] = v
+
+	// Return the params so calls can be chained.
+	return p
+}
 
 // Copy a list of params.
-func (p Params) Copy() Params {
-	dup := make(map[string]MaybeString, len(p))
-	for k, v := range p {
-		dup[k] = v
+func (p *Params) Copy() *Params {
+	dup := &Params{}
+	for k, v := range p.params {
+		dup.Add(k, v)
 	}
+
 	return dup
 }
 
+// Render params to a string.
+// Note that this does not escape special characters, this should already have been done before calling this method.
+func (p *Params) ToString(sep uint8) string {
+	var buffer bytes.Buffer
+	first := true
+
+	for k, v := range p.params {
+		if !first {
+			buffer.WriteString(fmt.Sprintf("%c", sep))
+		}
+		first = false
+
+		buffer.WriteString(fmt.Sprintf("%s", k))
+
+		switch v := v.(type) {
+		case String:
+			if strings.ContainsAny(v.String(), c_ABNF_WS) {
+				buffer.WriteString(fmt.Sprintf("=\"%s\"", v.String()))
+			} else {
+				buffer.WriteString(fmt.Sprintf("=%s", v.String()))
+			}
+		}
+	}
+
+	return buffer.String()
+}
+
+// Returns number of params.
+func (p *Params) Length() int {
+	if p == nil {
```

Comment:
You're checking for p==nil inconsistently; e.g. ToString will break on a nil Params but this won't. This won't be a problem if you make Params private and use an interface to access, so once you do that you can take the nil check out.